janitor/functions.py

```diff
@@ -4055,6 +4264,89 @@ def take_a_if_even(x):
     return df
 
 
+@pf.register_dataframe_method
+def join_conditional(
```

Comment:
@BaritoneBeard this is my own review; we can wait on @hectormz's as well, but I think the function here doesn't implement `join_conditional`, and its implementation should be changed to implement it correctly.

What's implemented in the function right now is instead a query operation, which is trivially done using `pandas`' existing `.query()` API. For example, one may do:

```python
df.query("left_column < right_column")
```

and we'll be done.

I think what's intended by `join_conditional` is filtering when joining two tables. Looking around online, I see a lot of discussion in the `pandas` dev chatter:

- https://github.com/pandas-dev/pandas/issues/8962
- https://github.com/pandas-dev/pandas/issues/10309

On studying those implementations, I think what's intended is the following implementation:

```python
@pf.register_dataframe_method
def join_conditional(df: pd.DataFrame, other: pd.DataFrame, conditions: List[str], join_kwargs={}):
    """Conditional join between one dataframe and another.

    This is syntactic sugar for joining (not merging) two dataframes and then filtering the results.

    The conditions that are passed in have to be compatible with dataframe query strings.
    For example, if you want to state that `one_column` must be less than `another_column`,
    then your condition string should be `"one_columns < another_column"`.

    Multiple conditions are allowed. 
    """

    joined_df = df.join(other, **join_kwargs)
    for condition in conditions:
        joined_df = joined_df.query(condition)
    return joined_df
```

Reply:
@ericmjl I agree. I see that my comments from last night didn't get posted. Whoops! But thanks for putting in the proposed change too

Reply:
Addressing @samukweku's point, self `join`s are supported by allowing `other` to be None in the edited implementation above.

**Action Items:**

1. Please change the implementation to match what is proposed above.
2. Please ensure the docstring is fully fleshed out like the other functions in the library. (I only put in the not-so-obvious points that a user might need to be reminded of.)